### PR TITLE
Initial thing that makes graphs

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This is the start of the chart-renderer node micro (ugh) service.

Right now this has two commands

`render` - Reads stdin json and produces an image on stdout. This will be used to generate graph images in your development environment.

`server` - Starts an express web server that has a single `POST /render` endpoint, which accepts a JSON body of series data, and responds with a png image.

Still TODO (doesn't need to be this PR)

* [x] Structured logging
* [x] validation of input (Maybe with json schema?)
* [x] Dockerfile
* [ ] More readme
* [ ] Write tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Sentry Chart Renderer
+
+This is a small service written in NodeJS to generate chart graphics using
+[node-canvas](https://www.npmjs.com/package/canvas)
++ [ECharts](https://echarts.apache.org/en/index.html). This service takes a
+configuration file defining styles, accepts JSON input for series data, and
+produces a png image to stdout.
+
+It can also opearte in a web-server mode to produce charts.
```
Comment (silent1mezzo) [inline README.md:9]:
```suggestion
It can also operate in a web-server mode to produce charts.
```

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  env: {node: true},
+  extends: ['sentry-app/strict'],
+  globals: {
+    require: false,
+    expect: false,
+    sinon: false,
+    MockApiClient: true,
+    TestStubs: true,
+    tick: true,
+    jest: true,
```
Comment (billyvg) [inline .eslintrc.js:11]: sinon still kickin around

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+function styleConfigLoader(path: string) {
+  const config = require(/* webpackIgnore: true */ path).default;
+
+  if (!validateStyleConfig(config)) {
+    throw new Error('Invalid style configuration');
+  }
+  return config;
+}
+
+yargsInit(process.argv.slice(2))
```
Comment (billyvg) [inline src/index.ts:24]: Would it make sense to have this in like `src/cli/` or something?

**Review by billyvg — COMMENTED** [review]
What is webpack being used for?

File: src/index.ts (hunk 3 above)
Comment (billyvg) [reply src/index.ts:24]: oh never mind, I see what it's doing now

Comment (EvanPurkhiser) [reply src/index.ts:24]: `yargs` is basically the parser / runner.

The whole thing is just one big cli tool really

File: src/renderStream.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import * as fs from 'fs';
+import * as util from 'util';
+
+import {renderSync} from './render';
+import {StyleConfig} from './types';
+import {validateRenderData} from './validate';
+
+const readAsync = util.promisify(fs.readFile);
```
Comment (billyvg) [inline src/renderStream.ts:8]: You could also `import * as fs from 'fs/promises'`

**Comment by EvanPurkhiser** [comment]
You're right I might be able to kill the webpack dependency.

It probably doesn't actually need it

**Comment by billyvg** [comment]
You can use `tsc` and get rid of babel + webpack

**Comment by EvanPurkhiser** [comment]
Yeah thinking that too now

File: package.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    "@babel/preset-typescript": "^7.13.0",
+    "babel-loader": "^8.2.2",
+    "eslint": "5.11.1",
+    "eslint-config-sentry-app": "1.50.0",
+    "file-loader": "^6.2.0",
+    "fork-ts-checker-webpack-plugin": "^6.1.0",
+    "jest": "^26.6.3",
+    "null-loader": "^4.0.1",
+    "prettier": "^2.2.1",
+    "ts-node": "^9.1.1",
```
Comment (dashed) [inline package.json:39]: Is this still used?

Comment (EvanPurkhiser) [reply package.json:39]: It was used to transpire webpack config. But after that’s gone it will be gone too

**Comment by dashed** [comment]
👍 on just relying on `tsc` unless there are specific features you need.

File: src/renderStream.ts (hunk 4 above)
Comment (EvanPurkhiser) [reply src/renderStream.ts:8]: Hmm, I tried this and it wasn't happy about getting `0` for the file descriptor

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,22 @@
+FROM node:12
```
Comment (dashed) [inline Dockerfile:1]: Would `node:14` be better? v14 is LTS, while v12 is in maintenance mode.

v14 could also enable `ES2020` as a compile target for tsc: https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping

Comment (EvanPurkhiser) [reply Dockerfile:1]: I picked 12 since it's what we were on in sentry. But maybe we could bump both 👀

File: src/renderStream.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    throw new Error('Invalid style key provided');
+  }
+
+  const [stream, dispose] = renderSync(style, data.series);
+
+  stream.pipe(process.stdout);
+
+  await new Promise((resolve, reject) => {
+    stream.on('end', resolve);
+    stream.on('error', reject);
```
Comment (dashed) [inline src/renderStream.ts:39]: Might want to `dispose()` on reject here. Otherwise line 42 would be unreachable on error.

Comment (EvanPurkhiser) [reply src/renderStream.ts:39]: good call

File: src/renderServer.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    const [stream, dispose] = renderSync(style, data.series);
+
+    response.status(200).contentType('png').attachment('chart.png');
+    stream.pipe(response);
+
+    await new Promise((resolve, reject) => {
+      stream.on('end', resolve);
+      stream.on('error', reject);
+    });
```
Comment (dashed) [inline src/renderServer.ts:45]: This seems awkward. I don't remember using streams like this in node.js.

File: src/renderStream.ts (hunk 7 above)
Comment (dashed) [reply src/renderStream.ts:39]: Should also apply for `renderSync()` as well.

File: src/renderStream.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import {StyleConfig} from './types';
+import {validateRenderData} from './validate';
+
+const readAsync = util.promisify(fs.readFile);
+
+/**
+ * Take JSON input from stdin and generate and output a graph to stdout
+ */
+export async function renderStream(styles: StyleConfig) {
+  const json = await readAsync(0, {encoding: 'utf8'});
```
Comment (dashed) [inline src/renderStream.ts:14]: There's a global `process.stdin.fd` which can point to the stdin fd depending the system. But it should be 0 in most cases.

Comment (EvanPurkhiser) [reply src/renderStream.ts:14]: that's much more clear for documentation reasons though. Thanks!

File: src/renderServer.ts (hunk 8 above)
Comment (EvanPurkhiser) [reply src/renderServer.ts:45]: Yeah there's probably some less awkward way to do this. It feels like node is still somewhat stuck in callback land

File: Dockerfile (hunk 6 above)
Comment (EvanPurkhiser) [reply Dockerfile:1]: Switched to node 14

File: src/renderStream.ts (hunk 4 above)
Comment (EvanPurkhiser) [reply src/renderStream.ts:8]: https://humanwhocodes.com/snippets/2019/05/nodejs-read-stream-promise/ uhg node

File: src/renderServer.ts (hunk 8 above)
Comment (dashed) [reply src/renderServer.ts:45]: It should be fine. It's pretty readable to reason about.

**Review by dashed — APPROVED** [review]
🚀

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import * as Sentry from '@sentry/node';
+import dotenv from 'dotenv';
+import yargsInit from 'yargs';
+
+import {renderServer} from './renderServer';
+import {renderStream} from './renderStream';
+import {resolveStylesConfig} from './stlyesLoader';
```
Comment (markstory) [inline src/index.ts:7]: Typo in styles.

Comment (EvanPurkhiser) [reply src/index.ts:7]: The real downside of typescript auto complete / imports

**Pull request merged**